<a href="https://colab.research.google.com/github/dSudario9/Decision_Tree/blob/main/Random_Forest_e_Gradiente_Boosting.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Base de Dados forneceida pela comunidade  de Medica para Estudos **texto em negrito**. **

**O Objetivo do projeto é encontrar o Modelo que melhor perfoma sobre o dataset , uma vez que precisamos de um profissional para entendermos melhor o dataset e . ** **texto em negrito**





In [ ]:
# Import das Bibliotecas bassicas 
import pandas as pd 


In [ ]:
# Import Dataset
df = pd.read_csv("/content/breast_cancer.csv")

In [ ]:
df.head() #Visualizando dados .

sample_id  mean_radius  ...  worst_fractal_dimension  diagnosis
0     842302        17.99  ...                  0.11890  malignant
1     842517        20.57  ...                  0.08902  malignant
2   84300903        19.69  ...                  0.08758  malignant
3   84348301        11.42  ...                  0.17300  malignant
4   84358402        20.29  ...                  0.07678  malignant

[5 rows x 32 columns]

**Criando Modelo**

In [ ]:
from sklearn.model_selection import train_test_split


In [ ]:
# Selecionando Variaveis Alvo 
X = df.drop(columns=["diagnosis"])
Y = df[["diagnosis"]]

In [ ]:
X.shape

(569, 31)

 **Separando Base para Train e Test ***

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)  # Usando apenas 20% da base.  

In [ ]:
print(X_train.shape)
print(X_test.shape) 
print(y_train.shape)
print(y_test.shape)

(455, 31)
(114, 31)
(455, 1)
(114, 1)


**Decision tree**

In [ ]:
# Biblioteca DecisionTreeClassifier

from sklearn.tree import DecisionTreeClassifier


In [ ]:
clf_dt = DecisionTreeClassifier()  #Classificador 

In [ ]:
clf_dt.fit(X_train, y_train)   # Trainando o Sistema

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [ ]:
#fazer predict 

y_pred = clf_dt.predict(X_test)

In [ ]:
# Calcular  acuraciá para nivel de comparação 
from sklearn.metrics import accuracy_score

accuracy_score(y_test, y_pred)*100

91.22807017543859

Como podemos vizualizar temos um bom resultado com acuraciá de 91.22807017543859,  porem vamos testar outros modelos e tentar melhorar o resultado ou fazermos comparações. 

**Modelo Random Fores**t

In [ ]:
# Biblioteca do Modelo de Classificação   RandomForestClassifier

from sklearn.ensemble import RandomForestClassifier

In [ ]:
# Classificador 

clf = RandomForestClassifier(n_estimators=250)

In [ ]:
# Treinando o  Modelo .

clf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=250,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [ ]:
# Fazer predição 
y_pred = clf.predict(X_test)


In [ ]:
# Acuraciá para comparação de resultados. 

accuracy_score(y_test, y_pred)*100

96.49122807017544

Com o modelo Random Forest nosso modelo ficou ainda melhor, pois a acuraciá foi superior ao modelo anterior onde chegamos ao resultado de 96.49122807017544 %. Nosso proximo passo é treinar um novo modelo para que possamos ter certeza de qual realmente é o melhor modelo . 

**Gradient Boosting**

In [ ]:
# Biblioteca do Modelo radientBoostingClassifier

from sklearn.ensemble import GradientBoostingClassifier

In [ ]:
# Classificador 

clf = GradientBoostingClassifier(n_estimators=250, subsample=1)

In [ ]:
# Treinando o Modelo GradientBoostingClassifier

clf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/ensemble/_gb.py:1454: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='deviance', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_impurity_split=None,
                           min_samples_leaf=1, min_samples_split=2,
                           min_weight_fraction_leaf=0.0, n_estimators=250,
                           n_iter_no_change=None, presort='deprecated',
                           random_state=None, subsample=1, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)

In [ ]:
# Predição 
y_pred = clf.predict(X_test)

In [ ]:
# Acuraciá para comparação de resultados.

accuracy_score(y_test, y_pred)*100

94.73684210526315

 O modelo GradientBoostingClassifier tambem tivemos um resultado satifatorio, porem ainda dar pra vizualizar que o modelo anterior os resultados foram melhores . Nesse modelo GradientBoostingClassifier a acuraciá foi de 94.73684210526315%.

**XGboost**

In [ ]:
# Biblioteca do Modelo XGboost 

from xgboost import XGBClassifier, train

In [ ]:
# Classificador do Modelo 

clf = XGBClassifier(n_estimators=200)

In [ ]:
# Treinando o Modelo .

clf.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:235: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.7/dist-packages/sklearn/preprocessing/_label.py:268: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=200, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
# Predição 

y_pred = clf.predict(X_test)

In [ ]:
# Calcular acuraciá 


print('Performace',accuracy_score(y_test, y_pred)*100)

Performace 98.24561403508771


 Como podemos ver o nosso modelo XGBoost mostrou nosso melhor modelo com acuraciá de 98.24561403508771%